File: src/main/scala/math/cat/PoSetMorphism.java
package math.cat;

import java.util.*;

/**
 * PoSetMorphism class: morphisms for posets.
 * All code is <a href="http://myjavatools.com/projects/Category/source.zip">here</a>
 */
public abstract class PoSetMorphism<X, Y> extends SetMorphism<X, PoSet<X>, Y, PoSet<Y>> {

  public PoSetMorphism(PoSet<X> domain, PoSet<Y> codomain) {
    super(domain, codomain);
    validate();
  }

  public PoSetMorphism(String id, PoSet<X> domain, PoSet<Y> codomain) {
    super(id, domain, codomain);
    validate();
  }

  private void validate() {
    for (X x1 : domain()) for (X x2 : domain()) {
      if (domain()._le_(x1, x2)) {
        assert codomain()._le_(apply(x1), apply(x2));
      }
    }
  }

  public String toString() {
    if (name() != null) return name();
    StringBuffer sb = new StringBuffer();
    for (X x : domain()) {
      if (sb.length() > 0) sb.append(", ");
      sb.append(x).append(" -> ").append(apply(x));
    }

    return "{" + sb + "}";
  }

  public abstract Y apply(X x);

  public <X, Y> PoSetMorphism<X, Y> PosetMorphism(final PoSet<X> domain, final PoSet<Y> codomain, final Map<X, Y> map) {
    for (X x : domain) {
      assert map.containsKey(x) : "Map should be defined on d0 element " + x;
    }

    return new PoSetMorphism<X, Y>(domain, codomain) {
      public Y apply(X x) {
        assert domain.contains(x) : "Argument not in d0";
        return map.get(x);
      }
    };
  }

  /**
   * Factory method. Builds unit morphism for a poset (identity function).
   *
   * @param s the poset
   * @return identity morphism on the given poset
   */
  public static <X> PoSetMorphism<X, X> unit(PoSet<X> s) {
    return new PoSetMorphism<X, X>(s, s) {
      public X apply(X x) {
        return x;
      }
    };
  }

  public static <
      X, // domain data type for the first morphism
      Y, // codomain data type for the first morphism (same as domain type for the second morphism)
      Z  // codomain data type for the first morphism
     >
  PoSetMorphism<X, Z> compose(
      final PoSetMorphism<X, Y> f,
      final PoSetMorphism<Y, Z> g
    ) {
    assert f.codomain().equals(g.domain()): "Composition not defined";
    return new PoSetMorphism<X, Z>(f.domain(), g.codomain()) {
      public Z apply(X x) {
        return g.apply(f.apply(x));
      }
    };
  }
}

File: src/main/scala/math/cat/PoSetMorphism.scala
package math.cat

import math.sets.PoSet
import scalakittens.Result
import scalakittens.Result.{OK, Outcome}


/**
  * Morphism for posets.
  * It is a kind of bad decision, to redefine composition here. Should use Scala cake pattern instead.
  * See http://scala.sygneca.com/patterns/component-mixins, http://lamp.epfl.ch/~odersky/papers/ScalableComponent.pdf
  */
class PoSetMorphism[X, Y] private(
  tag: String,
  override val d0: PoSet[X],
  override val d1: PoSet[Y],
  function: X => Y)
  extends SetMorphism[X, Y](tag, d0, d1, function) {
  type Domain = PoSet[X]
  type Codomain = PoSet[Y]

  def compose[Z](g: PoSetMorphism[Y, Z]): PoSetMorphism[X, Z] = {
    require(d1 == g.d0, "Composition not defined")
    new PoSetMorphism[X, Z]("a", d0, g.d1, (x: X) => g(this (x)))
  }
}

object PoSetMorphism {
  def check[X, Y](pom: PoSetMorphism[X, Y]): Result[PoSetMorphism[X, Y]] = {
    val r0 = SetMorphism.check[X, Y, PoSetMorphism[X, Y]](pom)
    val results = for {x <- pom.d0
                       y <- pom.d0
                       if pom.d0.le(x, y)
    } yield {
      val fx = pom(x)
      val fy = pom(y)
      Result.OKif(pom.d1.le(fx, fy),
        s"Since $x<=$y in domain, it should be $fx<=$fy in codomain")
    }
    r0 andThen Result.traverse(results) returning pom
  }

  def build[X, Y](tag: String, domain: PoSet[X], codomain: PoSet[Y], function: X => Y): Result[PoSetMorphism[X, Y]] =
    check(new PoSetMorphism[X, Y](tag, domain, codomain, function))

  def build[X, Y](domain: PoSet[X], codomain: PoSet[Y], function: X => Y): Result[PoSetMorphism[X, Y]] =
    check(new PoSetMorphism[X, Y]("", domain, codomain, function))

  def id[X](domain: PoSet[X]) =
    new PoSetMorphism[X, X]("1", domain, domain, x => x)

  def const[X, Y](domain: PoSet[X], codomain: PoSet[Y], value: Y) =
    new PoSetMorphism[X, Y](value.toString, domain, codomain, _ => value)
}
